### ISSUE TITLE
Revise config syntax detection?

### ISSUE THREAD

**Issue by simonsmith** [stylelint/stylelint#585]
Hey,

Bit confused over using `indentation` in 3.x

Given this file (with 4 spaces):

``` css
.test {
    color: red;
}
```

In `2.x` I would use the following config:

```js
{
  "rules": {
    "indentation": [2, 2]
  }
}
```

And get an error as expected:

```
node_modules/stylelint/dist/cli.js -f json test.css
[{"source":"/Users/simonsmith/Desktop/bug/test.css","errored":true,"warnings":[{"line":2,"column":5,"rule":"indentation","severity":2,"text":"Expected indentation of 2 spaces (indentation)"}]}]
```

Once upgraded to `3.x` I changed my config as recommended in [the documentation](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/configuration.md#severities-error--warning):

```
{
  "rules": {
    "indentation": 2
  }
}
```

This is the result:

```
[{"source":"/Users/simonsmith/Desktop/bug/test.css","deprecations":[{"text":"Numbered severities (0, 1, 2) have been deprecated, and in 4.0 they will be disabled. ","reference":"http://stylelint.io/?/docs/user-guide/configuration.md"}],"warnings":[{"text":"Invalid option value \"true\" for rule \"indentation\""}]}]
```

From what I can tell I'm not using a numbered severity, what am I doing wrong?

Just for extra info, if I switch it to `indentation: tab` it seems to report errors fine, and if I use `indentation: [2]` (also in the docs) I get an error:

```
RangeError: Maximum call stack size exceeded
    at transformRuleSettings (/Users/simonsmith/Desktop/bug/node_modules/stylelint/dist/buildConfig.js:137:79)
```

Any help is appreciated! Cheers

**Comment by jeddy3** [comment]
Hi Simon,

Thanks for the detailed report! Being the festive season it might take us a little while to get this one sorted, so I'm going to quickly write-up an explanation and workaround so you're not blocked :)

Because we intend to remove the numbered severities syntax in `4.0` we wanted to make the deprecation of it as seamless as possible in `3.0`, so we decided to automatically detect what syntax is in use, rather than have the user explicitly specify which syntax they are using (by the way of an option in their config e.g. `errorByDefault: true`).

What you've stumbled into is an edge-case that [we didn't think would come-up](https://github.com/stylelint/stylelint/pull/522#issuecomment-162384122), but, with the benefit of hindsight, it was a bad assumption. We assumed that if all the rules in the config start with a number, the config is using numbered severities. Now I realise that people might reach for `indentation` as their first rule, and so stylelint will incorrectly assume their minimal config is using numbered severities.

The problem will go away entirely in `4.0`, but I'm quite sure how to address it `3.0`. Hopefully @davidtheclark has some ideas?

To workaround the issue and to get on with using stylelint while we try to address the issue, you can add an _additional_ rule to your config that does _not_ use an `int` for its primary option i.e. everything but the `indentation` and the `-max-` rules. stylelint will then correctly detect your config is using the new config syntax, and the `2` option of `indentation` will be correctly processed.  

Or, if you're building configs for tests (e.g. integration ones in suit), you can use another non-`int` rule in the meantime.

I hope that helps for now.

**Comment by simonsmith** [comment]
Ah that's a helpful answer, thanks. This also explains my confusion around indentation working fine in the suit config, but not on its own. 

Cheers!

**Comment by davidtheclark** [comment]
I propose that we add a `numberedSeverities` property to the config, whose value is a boolean. If it's left out, the same auto-detection will happen as happens now (which has this one caveat). Otherwise, if set to `true` you'll force the numbered severities parsing; and if set to `false`, you'll force the new parsing. So to fix @simonsmith's problem you'd add `numberedSeverities: false` to the config. What do you think?

**Comment by simonsmith** [comment]
That could work. Maybe call it something like `legacySeverities` or `legacyNumberedSeverities` to make it clear it's on the way out?

Is the current system of auto-detection documented anywhere? It might be good to make a note of it, as the current docs show examples of just using `indentation` and it made it tricky to work out where I had gone wrong. Perhaps that could be added along with this fix.

**Comment by davidtheclark** [comment]
Released in 3.2.0
